### PR TITLE
build multiple base Images concurrently

### PR DESCRIPTION
**Pull request by louismurerwa**
This PR adds the ability for JIB to build multiple Images concurrently .

### CODE REVIEW

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java [hunk 1]
```diff
@@ -358,22 +358,27 @@ private void buildAndCacheApplicationLayers() {
...
-                    .call());
+            () -> {
+              List<Future<Image>> builtImages = new ArrayList<>();
+              for (Image image : results.baseImagesAndLayers.get().keySet()) {
+                Future<Image> builtImage =
+                    executorService.submit(
+                        () ->
+                            new BuildImageStep(
+                                    buildContext,
+                                    childProgressDispatcherFactory,
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java:371]: Just realized this. I actually think this code (and the code in `obtainBaseImageLayers` that we updated) will break once we have multiple `Images` due to `childProgressDispatcherFactory` (the progress reporting thing), which I haven't explained. Will need to fix it when the time comes. Will be easy to fix, so not a concern at all.

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java [hunk 2]
```diff
@@ -358,22 +358,27 @@ private void buildAndCacheApplicationLayers() {
...
+                                    image,
+                                    realizeFutures(
+                                        Verify.verifyNotNull(
+                                            results.baseImagesAndLayers.get().get(image))),
+                                    realizeFutures(Verify.verifyNotNull(results.applicationLayers)))
+                                .call());
+                builtImages.add(builtImage);
+              }
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java:379]: You can use `.entrySet()` to iterator over a Map. An entry is a (key, value) pair, so no need to call `Map.get(key)` to get the value.

```suggestion
              for (Map.Entry<Image, List<Future<PreparedLayer>>> entry :
                  results.baseImagesAndLayers.get().entrySet()) {
                Future<Image> builtImage =
                    executorService.submit(
                        () ->
                            new BuildImageStep(
                                    buildContext,
                                    childProgressDispatcherFactory,
                                    entry.getKey(), // base image
                                    realizeFutures(entry.getValue()), // base image layers
                                    realizeFutures(Verify.verifyNotNull(results.applicationLayers)))
                                .call());
                builtImages.add(builtImage);
              }
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java [hunk 3]
```diff
@@ -358,22 +358,27 @@ private void buildAndCacheApplicationLayers() {
   private void buildImage() {
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java:358]: Rename it to `buildImages()`? Also, rename other methods too (`pullBaseImage()` --> `pullBaseImages()`)?

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java (hunk 1 above)
Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java:371]: Sounds Good .I will be on the look out for the breaks.

**Pull request merged**